File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    answer := 0
    start := 0
    charMap := make(map[byte]int)
    c := []byte(s)

    for end := 0; end < len(c); end++ {
        if val, ok := charMap[c[end]]; ok {
            start = max(start, val + 1)
        }
        answer = max(answer, end - start + 1)
        charMap[c[end]] = end
    }
    return answer
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.c
int lengthOfLongestSubstring(char * s){
    int chars[128], idx = 128;
    while (idx--) chars[idx] = -1;

    int start = -1, length = 0, localLength;
    size_t c;
    while(s[++idx] != '\0') {
        c = (size_t) s[idx];
        if(chars[c] > start) start = chars[c];

        localLength = idx - start;
        if(length < localLength) length = localLength;
        chars[c] = idx; 
    }
    return (start > -1 ? length : idx);
}